File: app/src/main/java/com/ocs/sequre/app/base/BaseBottomSheet.java
package com.ocs.sequre.app.base;

import android.app.Dialog;
import android.content.Context;
import android.content.res.ColorStateList;
import android.graphics.Color;
import android.graphics.PorterDuff;
import android.os.Bundle;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.ViewParent;

import androidx.annotation.LayoutRes;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.coordinatorlayout.widget.CoordinatorLayout;
import androidx.lifecycle.ViewModelProvider;

import com.google.android.material.bottomsheet.BottomSheetBehavior;
import com.google.android.material.bottomsheet.BottomSheetDialogFragment;

import java.util.Objects;

import javax.inject.Inject;

import dagger.android.support.AndroidSupportInjection;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;

public abstract class BaseBottomSheet extends BottomSheetDialogFragment {

    private final CompositeDisposable disposables = new CompositeDisposable();
    protected BottomSheetBehavior<View> bottomSheetBehavior;

    @Inject
    protected ViewModelProvider.Factory factory;

    @Override
    public void onAttach(@NonNull Context context) {
        AndroidSupportInjection.inject(this);
        super.onAttach(context);
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(layoutRes(), container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        onViewBound(view);
        disposables.addAll(subscriptions());
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        View bottomSheet = ((View) Objects.requireNonNull(getView()).getParent());
        bottomSheet.setBackgroundTintMode(PorterDuff.Mode.CLEAR);
        bottomSheet.setBackgroundTintList(ColorStateList.valueOf(Color.TRANSPARENT));
        bottomSheet.setBackgroundColor(Color.TRANSPARENT);

        setHasOptionsMenu(true);
        setBottomSheetBehavior(getView());
    }

    @NonNull
    @Override
    public Dialog onCreateDialog(@Nullable Bundle savedInstanceState) {
        Dialog dialog = super.onCreateDialog(savedInstanceState);
        if (dialog.getWindow() != null) {
            dialog.getWindow().setGravity(Gravity.BOTTOM);
            dialog.getWindow().setBackgroundDrawableResource(android.R.color.transparent);
//            dialog.getWindow().clearFlags(WindowManager.LayoutParams.FLAG_DIM_BEHIND);
            dialog.setCancelable(false);
        }
        return dialog;
    }

    @Override
    public void onStart() {
        super.onStart();
        bottomSheetBehavior.setState(BottomSheetBehavior.STATE_EXPANDED);
    }

    private View getChild(ViewParent parent) {
        System.out.println("view = " + parent.getParent());
        if (parent.getParent() instanceof CoordinatorLayout) {
            return (View) parent;
        } else
            return getChild(parent.getParent());
    }

    private void setBottomSheetBehavior(View view) {
        bottomSheetBehavior = BottomSheetBehavior.from((View) view.getParent());
        bottomSheetBehavior.setPeekHeight(BottomSheetBehavior.PEEK_HEIGHT_AUTO, true);
        bottomSheetBehavior.addBottomSheetCallback(new BottomSheetBehavior.BottomSheetCallback() {

            @Override
            public void onStateChanged(@NonNull View bottomSheet, int newState) {
                if (BottomSheetBehavior.STATE_HIDDEN == newState) {
                    dismiss();
                }
            }

            @Override
            public void onSlide(@NonNull View bottomSheet, float slideOffset) {

            }
        });
    }

    @LayoutRes
    protected abstract int layoutRes();

    protected abstract void onViewBound(@NonNull View view);

    protected void addBottomSheetCallback(BottomSheetBehavior.BottomSheetCallback bottomSheetCallback) {
        bottomSheetBehavior.addBottomSheetCallback(bottomSheetCallback);
    }

    protected Disposable[] subscriptions() {
        return new Disposable[0];
    }

    protected void subscribe(Disposable d) {
        disposables.add(d);
    }

    @Override
    public void onDestroyView() {
        disposables.clear();
        super.onDestroyView();
    }
}

File: app/src/main/java/com/ocs/sequre/app/base/BaseBottomSheet.kt
package com.ocs.sequre.app.base

import android.view.View
import android.widget.Toast
import androidx.annotation.CallSuper
import androidx.appcompat.app.AlertDialog
import androidx.lifecycle.ViewModelProvider
import androidx.navigation.fragment.findNavController
import com.compact.app.CompactBottomSheetFragment
import com.compact.response.ApiException
import com.google.android.material.bottomsheet.BottomSheetBehavior
import com.google.android.material.snackbar.Snackbar
import com.google.android.material.textfield.TextInputLayout
import com.ocs.sequre.R
import com.ocs.sequre.data.remote.model.response.error.Error
import com.ocs.sequre.data.remote.model.response.error.ErrorStatus
import com.ocs.sequre.data.remote.model.response.error.ResponseError
import java.io.IOException
import javax.inject.Inject

abstract class BaseBottomSheet : CompactBottomSheetFragment() {
    @Inject
    lateinit var factory: ViewModelProvider.Factory
    private lateinit var progressBar: AlertDialog

    //    @NonNull
    //    @Override
    //    public Dialog onCreateDialog(@Nullable Bundle savedInstanceState) {
    //        Dialog dialog = super.onCreateDialog(savedInstanceState);
    //        if (dialog.getWindow() != null) {
    //            System.out.println("onCreateDialog");
    ////            dialog.getWindow().setGravity(Gravity.BOTTOM);
    ////            dialog.getWindow().setBackgroundDrawableResource(android.R.color.transparent);
    ////            dialog.getWindow().clearFlags(WindowManager.LayoutParams.FLAG_DIM_BEHIND);
    //            //dialog.setCancelable(false);
    //        }
    //        return dialog;
    //    }

    @CallSuper
    override fun onViewBound(view: View) {
        AlertDialog.Builder(view.context).apply {
            setView(R.layout.layout_progress_bar)
            setCancelable(false)
            create().run {
                window?.run {
                    setBackgroundDrawableResource(android.R.color.transparent)
                    allowEnterTransitionOverlap = true
                }
                progressBar = this
            }
        }
    }

    override fun onStart() {
        super.onStart()
        bottomSheetBehavior.state = BottomSheetBehavior.STATE_EXPANDED;
    }

    protected fun loading(it: Boolean) {
        if (it) progressBar.show() else progressBar.dismiss()
    }

    protected fun onSuccess(): () -> Unit {
        return {
            Toast.makeText(
                requireContext(),
                getString(R.string.saved_successfully),
                Toast.LENGTH_SHORT
            ).show()
            findNavController().navigateUp()
        }
    }

    protected fun onError(): (it: Throwable) -> Unit {
        return {
            it.printStackTrace()
            try {
                if (it is ApiException) {
                    if (it.code() >= 500) {
//                        onServerError()
                    } else {
                        it.error(ResponseError::class.java)?.run {
                            onApiException(code, errors)
                        }
                    }
                } else if (it is IOException) {
                    onIOException()
                } else {
                    onError(it.message!!)
                }
            } catch (e: Exception) {
                onIOException()
            }
        }
    }

    private fun onApiException(code: ErrorStatus, errors: List<Error>) {
        if (code == ErrorStatus.VALIDATION) {
            for (e in errors) {
                requireView().findViewWithTag<TextInputLayout>(e.path)
                    ?.apply {
                        error = e.message
                        setEndIconOnClickListener { performClick() }
                    }
            }
        } else if (errors.isNotEmpty() && code != ErrorStatus.VALIDATION) {
            Toast.makeText(context, errors[0].message, Toast.LENGTH_LONG).show()
        }
    }

    private fun onServerError() {
        Snackbar.make(
            requireView(),
            "Internal Server Error",
            Snackbar.LENGTH_LONG
        ).show()
    }

    private fun onIOException() {
        Snackbar.make(
            requireView(),
            "Internet connection lost",
            Snackbar.LENGTH_LONG
        ).show()
    }

    private fun onError(message: String) {
        Toast.makeText(context, message, Toast.LENGTH_LONG).show()
    }
}